src/boot/efi/util.h

```diff
@@ -43,6 +43,18 @@ static inline void freep(void *p) {
 
 #define _cleanup_free_ _cleanup_(freep)
 
+static inline void erase_obj(void *p) {
+        size_t l;
+        if (!p)
+                return;
+        l = __builtin_object_size(p, 0);
```

Comment:
hmm, this fields a bit unsafe, given that the compiler is free to return SIZE_MAX from that builtin whenever it likes. and iirc it actually does that quite often, in particular when optimization is off..

Reply:
I manually checked that it always works for stack buffers like this. It's also how it's done elsewhere in systemd userspace. And there's the assertion to catch it if it breaks. And those assertions will always trigger since it's cleanup code. So I think we're good.

Reply:
hmm, we use this __builtin_object_size() only on heap memory so far, and there is just a secondary size, the primary one is malloc_usable_size() which is guaranteed to work. We only use it btw, since if _FORTIFY_SOURCES is turned on we'll end up aborting if we touch more than __builtin_object_size() bytes even if malloc_usable_size() claims more.

Reply:
hmm, so, I would feel more comfortable if we could trigger a build-time error if this doesn't work.

maybe define:

```c
#define _cleanup_erase_(type) \
        assert_cc(__builtin_object_size(type, 0) != SIZE_MAX); \
        _cleanup_(erase_object) type
```

or so.

Then we'd use it like so:

```c
_cleanup_erase_(struct sha256_ctx) hash;
```

which is just a () away from your current version?

(or alternatively, maybe the assert_cc() can just be inside of the cleanup function?)

Reply:
btw, according to this https://man.netbsd.org/__builtin_object_size.3  the concept exists since gcc 4.1 (which is fine by us, we currently have a minimum requirement of gcc 4.7 according to README). But it says (as I mentioned earlier) it doesn't work at -O0 on "current" gcc.  Which would be quite a bummer. Now, not sure what "current" was at time of writing, maybe actually 2022 gcc is fine with it, but I wonder since when?

Reply:
I had tried the assert_cc from inside the function before, but it didn't work. Doing it in the macro sounds like a good compromise, however.

Reply:
Hmmm that doesn't work as nicely for arrays though...

Reply:
Okay I have a pretty cool solution that does a compile time assertion inside of the cleanup function using later-stage evaluation. Works well. Same trick as what the Linux kernel uses.